### PR TITLE
KAFKA-17614: Remove AclAuthorizer

### PR DESCRIPTION
**Pull request by FrankYang0529**
Remove AclAuthorizer as it has been superseded by StandardAuthorizer

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 1]
```diff
@@ -710,7 +710,7 @@ class AuthorizerIntegrationTest extends AbstractAuthorizerIntegrationTest {
   }
 
   @ParameterizedTest
-  @ValueSource(strings = Array("zk", "kraft"))
+  @ValueSource(strings = Array("kraft"))
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:647]: Can we also update the `sendRequest()` method just above and remove the ZK/KRaft logic?

File: core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationTest.scala [hunk 2]
```diff
@@ -107,7 +107,7 @@ class DelegationTokenEndToEndAuthorizationTest extends EndToEndAuthorizationTest
   }
 
   @ParameterizedTest
-  @ValueSource(strings = Array("kraft", "zk"))
+  @ValueSource(strings = Array("kraft"))
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationTest.scala:103]: Let's also update the `configureSecurityBeforeServersStart()` method and remove the ZK logic

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala [hunk 3]
```diff
@@ -39,7 +38,7 @@ import org.apache.kafka.common.resource.PatternType.{LITERAL, PREFIXED}
 import org.apache.kafka.common.security.auth._
 import org.apache.kafka.coordinator.group.GroupCoordinatorConfig
 import org.apache.kafka.security.authorizer.AclEntry.WILDCARD_HOST
-import org.apache.kafka.server.config.{ServerConfigs, ReplicationConfigs, ServerLogConfigs, ZkConfigs}
+import org.apache.kafka.server.config.{ServerConfigs, ReplicationConfigs, ServerLogConfigs}
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:41]: Can we update the javadoc of this file and remove the ZooKeeper mention?

File: core/src/test/scala/integration/kafka/api/PlaintextEndToEndAuthorizationTest.scala [hunk 4]
```diff
@@ -89,7 +89,7 @@ class PlaintextEndToEndAuthorizationTest extends EndToEndAuthorizationTest {
   }
 
   @ParameterizedTest
-  @ValueSource(strings = Array("kraft", "zk"))
+  @ValueSource(strings = Array("kraft"))
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/PlaintextEndToEndAuthorizationTest.scala:92]: In the `setUp()` method above we can change `ZkSasl` to `KafkaSasl`

File: docs/security.html [hunk 5]
```diff
@@ -1277,8 +1277,6 @@ <h3 class="anchor-heading"><a id="security_authz" class="anchor-link"></a><a hre
     Configured implementations must extend <code>org.apache.kafka.server.authorizer.Authorizer</code>.
     Kafka provides default implementations which store ACLs in the cluster metadata (either Zookeeper or the KRaft metadata log).
 
-    For Zookeeper-based clusters, the provided implementation is configured as follows:
```
Comment (mimaison) [inline docs/security.html:1280]: Let's also remove the ZooKeeper mention on the line above since Kafka does not provide a ZooKeeper implementation anymore.

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala [hunk 6]
```diff
@@ -10,11 +10,10 @@
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
- **/
+ * */
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala]: Is this change intended?

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala [hunk 7]
```diff
@@ -87,11 +87,7 @@ class RequestQuotaTest extends BaseRequestTest {
     properties.put(GroupCoordinatorConfig.GROUP_INITIAL_REBALANCE_DELAY_MS_CONFIG, "0")
     properties.put(BrokerSecurityConfigs.PRINCIPAL_BUILDER_CLASS_CONFIG, classOf[RequestQuotaTest.TestPrincipalBuilder].getName)
     properties.put(ServerConfigs.UNSTABLE_API_VERSIONS_ENABLE_CONFIG, "true")
-    if (isKRaftTest()) {
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:90]: A bunch of other methods (`clientActions()`, `clusterActions()`, etc) still have KRaft/ZooKeeper conditional code, can we remove that as well?

File: core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala [hunk 8]
```diff
@@ -1,210 +0,0 @@
-/**
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala]: Do we have another test that validates this with KRaft?

File: core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala [hunk 9]
```diff
@@ -1,169 +0,0 @@
-/**
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala:1]: Why are we deleting this test? Is this covered by another test class?

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala [hunk 10]
```diff
@@ -392,161 +409,17 @@ class AuthorizerTest extends QuorumTestHarness with BaseAuthorizerTest {
...
     removeAcls(authorizer1, Set.empty, resource)
     TestUtils.waitAndVerifyAcls(Set.empty[AccessControlEntry], authorizer1, resource)
-    if (quorum.equals(ZK)) {
-      assertFalse(zkClient.resourceExists(resource))
-    }
 
     //test removing last acl also deletes ZooKeeper path
     acls = changeAclAndVerify(Set.empty, Set(acl1), Set.empty)
     changeAclAndVerify(acls, Set.empty, acls)
-    if (quorum.equals(ZK)) {
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:402]: Can we update the comment above that still mentions ZooKeeper?

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala [hunk 11]
```diff
@@ -142,16 +161,14 @@ class AuthorizerTest extends QuorumTestHarness with BaseAuthorizerTest {
 
   // Authorizing the empty resource is not supported because we create a znode with the resource name.
   @ParameterizedTest
-  @ValueSource(strings = Array(KRAFT, ZK))
+  @ValueSource(strings = Array(KRAFT))
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:122]: Also update the comment above that mentions `znode`

File: core/src/test/scala/unit/kafka/security/authorizer/BaseAuthorizerTest.scala [hunk 12]
```diff
@@ -1,375 +0,0 @@
...
-  val superUsers = "User:superuser1; User:superuser2"
-  val username = "alice"
-  val principal = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, username)
-  val requestContext: RequestContext = newRequestContext(principal, InetAddress.getByName("192.168.0.1"))
-  val superUserName = "superuser1"
-  var config: KafkaConfig = _
-  var zooKeeperClient: ZooKeeperClient = _
-  var resource: ResourcePattern = _
-
-  @Test
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/security/authorizer/BaseAuthorizerTest.scala]: What about these tests? Are these tested somewhere else?

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala [hunk 13]
```diff
@@ -392,161 +409,17 @@ class AuthorizerTest extends QuorumTestHarness with BaseAuthorizerTest {
...
-      future.get(10, TimeUnit.SECONDS)
-
-      assertEquals(acls, getAcls(aclAuthorizer3, resource))
-    } finally {
-      aclAuthorizer3.close()
-      executor.shutdownNow()
-    }
-  }
-
-  @Test
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:466]: Are these tests not relevant in KRaft mode?

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala [hunk 14]
```diff
@@ -786,291 +641,17 @@ class AuthorizerTest extends QuorumTestHarness with BaseAuthorizerTest {
...
-    ZkConfigs.ZK_SSL_CONFIG_TO_SYSTEM_PROPERTY_MAP.asScala.keys.foreach(prop => prop match {
-      case ZkConfigs.ZK_SSL_CLIENT_ENABLE_CONFIG | ZkConfigs.ZK_SSL_CRL_ENABLE_CONFIG | ZkConfigs.ZK_SSL_OCSP_ENABLE_CONFIG =>
-        assertEquals("true", KafkaConfig.zooKeeperClientProperty(zkClientConfig, prop).getOrElse("<None>"))
-      case ZkConfigs.ZK_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG =>
-        assertEquals("false", KafkaConfig.zooKeeperClientProperty(zkClientConfig, prop).getOrElse("<None>"))
-      case _ => assertEquals(prefixedValue, KafkaConfig.zooKeeperClientProperty(zkClientConfig, prop).getOrElse("<None>"))
-    })
-  }
-
-  @Test
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:993]: Are these tests not relevant in KRaft mode?

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. I left a few comments.

Also `AclAuthorizer` is still used in `tests/kafkatest/services/kafka.py`. There's also a mention in `docker/native/reflect-config.json`.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 1 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:647]: Remove `isKRaftTest` and related comment in `sendRequests`.

File: core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationTest.scala (hunk 2 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/DelegationTokenEndToEndAuthorizationTest.scala:103]: Removed it. Thank you.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala (hunk 3 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:41]: Yes, removed related ZooKeeper mention.

File: core/src/test/scala/integration/kafka/api/PlaintextEndToEndAuthorizationTest.scala (hunk 4 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/PlaintextEndToEndAuthorizationTest.scala:92]: Updated it. Thanks.

File: docs/security.html (hunk 5 above)
Comment (FrankYang0529) [reply docs/security.html:1280]: Remove `Zookeeper`.

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala (hunk 6 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala]: No, it's not intended. Reverted it. Thanks.

File: core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala (hunk 7 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/server/RequestQuotaTest.scala:90]: Yes, remove `isKRaftTest` in other methods.

File: core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala (hunk 8 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala]: No, we don't have. I disable this test now. I prefer to create another Jira for followup. What do you think?

File: core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala (hunk 9 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/SaslPlainSslEndToEndAuthorizationTest.scala:1]: The only test case `testAcls` uses `zkClient` directly. Do we need to update to kraft? Thanks.

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala (hunk 10 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:402]: Yes, removed it.

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala (hunk 11 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:122]: Updated it.

File: core/src/test/scala/unit/kafka/security/authorizer/BaseAuthorizerTest.scala (hunk 12 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/security/authorizer/BaseAuthorizerTest.scala]: Reverted this change. Thanks for catching this.

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala (hunk 13 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:466]: Keep `testLocalConcurrentModificationOfResourceAcls`, but still remove other test cases. It looks like `StandardAuthorizer` is not like `AclAuthorizer`. It can't listen change on other `Authorizer` object, so we can't run test cases like `testDistributedConcurrentModificationOfResourceAcls`.

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala (hunk 14 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:993]: This case overrides `AclAuthorizer#processAclChangeNotification`. IIRC, we don't have similar mechanism in `StandardAuthorizer`.

File: docs/security.html [hunk 15]
```diff
@@ -1275,10 +1275,8 @@ <h3 class="anchor-heading"><a id="security_sasl" class="anchor-link"></a><a href
     <h3 class="anchor-heading"><a id="security_authz" class="anchor-link"></a><a href="#security_authz">7.5 Authorization and ACLs</a></h3>
     Kafka ships with a pluggable authorization framework, which is configured with the <code>authorizer.class.name</code> property in the server configuration.
     Configured implementations must extend <code>org.apache.kafka.server.authorizer.Authorizer</code>.
-    Kafka provides default implementations which store ACLs in the cluster metadata (either Zookeeper or the KRaft metadata log).
+    Kafka provides default implementations which store ACLs in the cluster metadata (KRaft metadata log).
```
Comment (mimaison) [inline docs/security.html:1278]: `default implementations` -> `a default implementation`, since there's just one now

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala [hunk 16]
```diff
@@ -786,291 +614,17 @@ class AuthorizerTest extends QuorumTestHarness with BaseAuthorizerTest {
...
 
-  @Test
-  def testThrowsOnAddPrefixedAclIfInterBrokerProtocolVersionTooLow(): Unit = {
-    givenAuthorizerWithProtocolVersion(Option(IBP_2_0_IV0))
-    val e = assertThrows(classOf[ApiException],
-      () => addAcls(authorizer1, Set(denyReadAcl), new ResourcePattern(TOPIC, "z_other", PREFIXED)))
-    assertTrue(e.getCause.isInstanceOf[UnsupportedVersionException], s"Unexpected exception $e")
-  }
-
-  @Test
```
Comment (mimaison) [inline core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:807]: Is this test not relevant anymore?

File: docs/security.html (hunk 15 above)
Comment (mimaison) [inline docs/security.html]: `default implementations` -> `a default implementation`, since there's just one now

File: core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala [hunk 17]
```diff
@@ -74,12 +74,13 @@ object DescribeAuthorizedOperationsTest {
   }
 }
 
+@Disabled("TODO: change to kraft")
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala]: Have we opened a Jira for this?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 18]
```diff
@@ -34,24 +34,20 @@ import org.apache.kafka.common.message.CreatePartitionsRequestData.CreatePartiti
...
 import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 import org.apache.kafka.common.record.{MemoryRecords, RecordBatch, SimpleRecord}
 import org.apache.kafka.common.requests.OffsetFetchResponse.PartitionData
 import org.apache.kafka.common.requests._
 import org.apache.kafka.common.resource.PatternType.{LITERAL, PREFIXED}
 import org.apache.kafka.common.resource.ResourceType._
 import org.apache.kafka.common.resource.{PatternType, Resource, ResourcePattern, ResourcePatternFilter, ResourceType}
-import org.apache.kafka.common.security.auth.{KafkaPrincipal, SecurityProtocol}
+import org.apache.kafka.common.security.auth.{KafkaPrincipal}
```
Comment (mimaison) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: We can remove the curly brackets

**Review by mimaison — COMMENTED** [review]
Thanks for the updates, I left a few more comments.

File: core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala (hunk 16 above)
Comment (FrankYang0529) [reply core/src/test/scala/unit/kafka/security/authorizer/AuthorizerTest.scala:807]: Hi @mimaison, thanks for catching this. In `AuthorizerTest`, `StandardAuthorizer` uses `MockAclMutator`, so this test can't pass. I tried to migrate the case to `AuthorizerIntegrationTest`, but the implementation of validating acl is different between `AclAuthorizer#validateAclBinding` and `AclControlManager#validateNewAcl`. Do you think that we should add `/` check for it? Thanks.

https://github.com/apache/kafka/blob/b917b85e62373ca66ef249c86b102b5ef8ee7195/core/src/main/scala/kafka/security/authorizer/AclAuthorizer.scala#L122-L127

https://github.com/apache/kafka/blob/b917b85e62373ca66ef249c86b102b5ef8ee7195/metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java#L127-L163

File: core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala (hunk 17 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/DescribeAuthorizedOperationsTest.scala]: Create a new Jira for it https://issues.apache.org/jira/browse/KAFKA-17833.

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 18 above)
Comment (FrankYang0529) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Removed it. Thanks.

File: docs/security.html (hunk 15 above)
Comment (FrankYang0529) [reply docs/security.html]: Thanks. Updated it.

**Review by mimaison — APPROVED** [review]
LGTM
